service/raft/group0_state_machine.cc

```diff
@@ -43,11 +43,19 @@
 #include "db/config.hh"
 #include "replica/database.hh"
 #include "service/raft/group0_state_machine_merger.hh"
+#include "gms/feature_service.hh"
 
 namespace service {
 
 static logging::logger slogger("group0_raft_sm");
 
+group0_state_machine::group0_state_machine(raft_group0_client& client, migration_manager& mm, storage_proxy& sp, storage_service& ss, raft_address_map& address_map, gms::feature_service& feat, bool topology_change_enabled)
+        : _client(client), _mm(mm), _sp(sp), _ss(ss), _address_map(address_map), _topology_change_enabled(topology_change_enabled)
+        , _topology_on_raft_support_listener(feat.supports_consistent_topology_changes.when_enabled([this] () noexcept {
+            _topology_change_enabled = true;
```

Comment:
But by the time the first pull may be triggered, already requiring topology to be included --- let's say that the topology coordinator has committed the first group 0 command as part of the upgrade procedure --- we might still not have noticed that the feature is enabled, no? In which case we'll skip pulling topology, and that's bad.

Reply:
Hmm, right. Rookie mistake from my part - assuming that the feature is enabled on other nodes if it is enabled on the local node. Before we start the upgrade, we need to somehow make sure that other nodes will, from that moment on, pull the topology snapshot...

Reply:
In the end, it will be the responsibility of the administrator to wait until the feature becomes enabled on all nodes. I mention this in the `topology-over-raft.md` doc and added a comment in the code about the improper use of the cluster features.

Reply:
Please also add a comment in the documentation issue:
https://github.com/scylladb/scylladb/issues/15934